src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+    image. <gdImageCreateFromAvif> returns a <gdImagePtr> to the new
+    truecolor image, or NULL if unable to load the image (most often
+    because the file is corrupt or does not contain a AVIF
+    image). <gdImageCreateFromAvif> does not close the file.
+
+
+    This function creates a gdIOCtx struct from the file pointer it's passed.
+    And then it relies on <gdImageCreateFromAvifCtx> to do the real decoding work.
+    If the file contains an image sequence, we simply read the first one, discarding the rest.
+
+  Variants:
+
+    <gdImageCreateFromAvifPtr> creates an image from AVIF data
+    already in memory.
+
+    <gdImageCreateFromAvifCtx> reads data from the function
+    pointers in a <gdIOCtx> structure.
+
+  Parameters:
+
+    infile - pointer to the input file
+
+  Returns:
+
+    A pointer to the new truecolor image.  This will need to be
+    destroyed with <gdImageDestroy> once it is no longer needed.
+
+    On error, returns 0.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvif (FILE * infile)
+{
+  gdImagePtr image;
+  gdIOCtx *ctx = gdNewFileCtx(infile);
+
+  if (!ctx) {
+    return GD_FALSE;
+  }
+
+  image = gdImageCreateFromAvifCtx(ctx);
+  // ctx->gd_free(ctx);
+
+  return image;
+}
+
+/*
+  Function: gdImageCreateFromAvifPtr
+
+    See <gdImageCreateFromAvif>.
+
+  Parameters:
+
+    size            - size of Avif data in bytes.
+    data            - pointer to Avif data.
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifPtr (int size, void *data)
+{
+  gdImagePtr image;
+  gdIOCtx *ctx = gdNewDynamicCtxEx(size, data, 0);
+
+  if (!ctx)
+    return 0;
+
+  image = gdImageCreateFromAvifCtx(ctx);
+  ctx->gd_free(ctx);
+
+  return image;
+}
+
+/*
+  Function: gdImageCreateFromAvifCtx
+
+    See <gdImageCreateFromAvif>.
+    
+    Additional details: the AVIF library comes with functions to create an IO object from
+    a file and from a memory pointer. Of course, it doesn't have a way to create an IO object
+    from a gdIOCtx. So, here, we use our own helper function, createAvifIOfromCtx().
+
+    Otherwise, we create the image by calling AVIF library functions in order:
+    * avifDecoderCreate(), to create the decoder
+    * avifDecoderSetIO(), to tell libavif how to read from our data structure
+    * avifDecoderParse(), to parse the image
+    * avifDecoderNextImage(), to read the first image from the decoder
+    * avifRGBImageSetDefaults(), to create the avifRGBImage
+    * avifRGBImageAllocatePixels(), to allocate memory for the pixels
+    * avifImageYUVToRGB(), to convert YUV to RGB
+    
+    Finally, we create a new gd image and copy over the pixel data.
+  
+  Parameters:
+
+    ctx              - a gdIOCtx struct
+*/
+BGD_DECLARE(gdImagePtr) gdImageCreateFromAvifCtx (gdIOCtx * ctx)
+{
+  int x, y;
+  uint8_t *p8; 
+  uint16_t *p16;
+  gdImage *im = NULL;
+  avifResult result;
+  avifIO *io;
+  avifDecoder *decoder;
+  avifRGBImage rgb;
+
+  decoder = avifDecoderCreate();
```

Comment:
do any of these avif calls need their own error checking ?  i.e. if any of them fail, you need to check & return errors accordingly.

Reply:
Some do! This one doesn't; if it can't `malloc`, it `abort`s.

Checked the rest of my avif calls, and for the ones that can return an error code, I've already got a check. The others don't return anything.